homeassistant/components/logbook/__init__.py

```diff
@@ -209,12 +216,62 @@ def _async_describe_event(
     platform.async_describe_events(hass, _async_describe_event)
 
 
+def _async_determine_event_types(
+    hass: HomeAssistant, entity_ids: list[str] | None, device_ids: list[str] | None
+) -> tuple[str, ...]:
+    """Reduce the event types based on the entity ids and device ids."""
+    external_events: dict[
+        str, tuple[str, Callable[[LazyEventPartialState], dict[str, Any]]]
+    ] = hass.data.get(DOMAIN, {})
+    if not entity_ids and not device_ids:
+        return (*ALL_EVENT_TYPES_EXCEPT_STATE_CHANGED, *external_events)
+    config_entry_ids: set[str] = set()
+    intrested_event_types: set[str] = set()
+
+    if entity_ids:
+        #
+        # Home Assistant doesn't allow firing events from
+        # entities (at least in code review)
+        # so we have a limited list to check
+        #
+        # automations and scripts can refer to entities
+        # but they do not have a config entry so we need
+        # to add them.
+        #
+        # We also allow entity_ids to be recorded via
+        # manual logbook entries.
+        #
+        intrested_event_types |= ENTITY_EVENTS_WITHOUT_CONFIG_ENTRY
+
+    if device_ids:
+        dev_reg = dr.async_get(hass)
+        for device_id in device_ids:
+            if (device := dev_reg.async_get(device_id)) and device.config_entries:
+                config_entry_ids |= device.config_entries
+        intrested_domains: set[str] = set()
+        for entry_id in config_entry_ids:
+            if entry := hass.config_entries.async_get_entry(entry_id):
+                intrested_domains.add(entry.domain)
+        for external_event, domain_call in external_events.items():
+            if domain_call[0] in intrested_domains:
+                intrested_event_types.add(external_event)
```

Comment:
We may find we need to refactor this later to store events by domain when they are registered if we ever get more than 100 of them.